### PR TITLE
Make ExchangeService more or less thread safe

### PR DESCRIPTION
**Pull request by candrews**
HttpClientContext isn't thread safe, and isn't meant for reuse, so create a new one each time it's needed
Use a shared cookied store per ExchangeService instance, clearing when necessary
ExchangeService is now thread safe, as long as certain properties (meant to be only set once at initialization time) are not changed oftentime while ExchangeService is being used
Safely manage HttpClient instances so they're lazy initialized used double checked locking
Made fields that can be final final, which improve understability and encourages thread safe design

Resolves gh-371

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@candrews__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java [hunk 1]
```diff
@@ -132,24 +134,26 @@
...
-
-  private Map<String, String> httpResponseHeaders = new HashMap<String, String>();
+  private Map<String, String> httpHeaders = new ConcurrentHashMap<String, String>();
 
   private WebProxy webProxy;
 
-  protected CloseableHttpClient httpClient;
+  private volatile CloseableHttpClient httpClient;
+
+  private final Object httpClientLock = new Object();
```
Comment (serious6) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: why you didnt use [ReentrantLocks](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReentrantLock.html)?

**Comment by serious6** [comment]
@candrews thanks for your contribution. I think this needs a little more review. Any chance of adding some tests to prove thread-safety?

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java (hunk 1 above)
Comment (candrews) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: Because synchronized keyword locking is reentrant. If we wanted to avoid use of the synchronized keyword and implement the locking manually then ReentrantLock, as you said, would be the way to go... But I don't see the need for that complexity (the keyword approach is more readable).

**Comment by candrews** [comment]
What do you envision such a test to look like / how would you like it to work? (I don't think there is a way to prove thread safety in unit or integration tests - we can do some concurrency testing but it wouldn't be a proof).

BTW, the double check locking implementation is textbook - there's no creativity/originality there so I don't believe it to be risky or flawed in any way. For the code that this essentially is a copy of, see https://en.m.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java

I'm very eager to get this fix in as the library is more or less unusable without it in any meaningful way, so please help me in getting this merged :-)

Also, it's unfortunate that this locking in necessary at all. If ExchangeService was refactored to be immutable and have its instances created using a builder pattern, then no locking would be necessary improving runtime performance and simplicity beyond what the current design allows. Would you be open to such an API change?

**Comment by serious6** [comment]
As this api is a simple portation form c# to java I am completely open to anything which improves the design and usability of it. We also had lots of PR that where already targeting these kind of issues. Nevertheless I know that there are plenty of places that are far away from good design. 
If you like to provide any contribution in that area you are welcome and I think others will support too.

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java [hunk 2]
```diff
@@ -888,15 +893,14 @@ private void saveHttpResponseHeaders(Map<String, String> headers) {
...
-      if (ExchangeServiceBase.binarySecret == null) {
-        Random randomNumberGenerator = new Random();
-        ExchangeServiceBase.binarySecret = new byte[256 / 8];
-        randomNumberGenerator.nextBytes(binarySecret);
-      }
-
-      return ExchangeServiceBase.binarySecret;
-    }
+	Random randomNumberGenerator = new SecureRandom();
+	final byte[] bytes = new byte[256 / 8];
```
Comment (serious6) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: while refactoring please avoid magic numbers.

Comment (candrews) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: As you noted, I only refactored this magic number containing code; I don't know the significance of the magic numbers (256 and 8). Do you know or do you know how I can find out so I make them into appropriately named constants?

I see these lines were added in https://github.com/OfficeDev/ews-java-api/blame/45e3fe6301e5824a1186e9e279200df5d57d73fe/src/main/java/microsoft/exchange/webservices/data/ExchangeServiceBase.java but the commit comment ("Initial Commit") isn't helpful, and I don't know where else to look.

**Comment by candrews** [comment]
I'll look into implementing a builder style ExchangeService - thanks for the info. In the mean time, is this something you could merge?

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java (hunk 2 above)
Comment (vboctor) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: Since this is an existing issue and is unrelated, we should keep it as is.  If you desire to fix it, then do it as a separate PR.  Let's keep PRs focused on one issue.

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java [hunk 3]
```diff
@@ -230,6 +208,9 @@ private void initializeHttpPoolingClient() {
   public void setMaximumPoolingConnections(int maximumPoolingConnections) {
     if (maximumPoolingConnections < 1)
       throw new IllegalArgumentException("maximumPoolingConnections must be 1 or greater");
+    if(httpPoolingClient!=null){
```
Comment (vboctor) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: Shouldn't we have space after the "if" and around the != and after the ")"?  "if (httpPoolingClient != null) {"

File: src/main/java/microsoft/exchange/webservices/data/core/WebProxy.java [hunk 4]
```diff
@@ -45,8 +45,7 @@
    * @param port proxy port.
    */
   public WebProxy(String host, int port) {
-    this.host = host;
-    this.port = port;
+    this(host,port,null);
```
Comment (vboctor) [inline src/main/java/microsoft/exchange/webservices/data/core/WebProxy.java]: Include a space after commas

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java [hunk 5]
```diff
@@ -132,24 +134,26 @@
...
 
-  private Map<String, String> httpHeaders = new HashMap<String, String>();
-
-  private Map<String, String> httpResponseHeaders = new HashMap<String, String>();
+  private Map<String, String> httpHeaders = new ConcurrentHashMap<String, String>();
 
   private WebProxy webProxy;
 
-  protected CloseableHttpClient httpClient;
+  private volatile CloseableHttpClient httpClient;
```
Comment (vboctor) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: Do we really need volatile here?  Doesn't the use of synchronized between the two if checks does the trick of creating a barrier that makes the compiler assume that non local variables may be changed by another thread?

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java [hunk 6]
```diff
@@ -132,24 +134,26 @@
...
 
   /**
    * The server info.
    */
   private ExchangeServerInfo serverInfo;
 
-  private Map<String, String> httpHeaders = new HashMap<String, String>();
-
-  private Map<String, String> httpResponseHeaders = new HashMap<String, String>();
+  private Map<String, String> httpHeaders = new ConcurrentHashMap<String, String>();
```
Comment (vboctor) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: Why do we need concurrent hash map for instance variables on ExchangeService?  Are we designing with the assumption that:
- A single instance of ExchangeService can be used by a single server.
- Multiple threads can run in parallel each using its own instance of ExchangeService.

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java [hunk 7]
```diff
@@ -252,29 +233,37 @@ public void setMaximumPoolingConnections(int maximumPoolingConnections) {
...
-      } catch (IOException e) {
-        LOG.debug(e);
+      synchronized(httpPoolingClientLock){
+    	if (httpPoolingClient != null) {
+          try {
+            httpPoolingClient.close();
+          } catch (IOException e) {
+            LOG.debug(e);
+          }
+          httpPoolingClient=null;
```
Comment (vboctor) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: Add spaces around the equal sign.

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java [hunk 8]
```diff
@@ -383,8 +370,6 @@ private void prepareHttpWebRequestForUrl(URI url, boolean acceptGzipEncoding, bo
     prepareCredentials(request);
 
     request.prepareConnection();
-
-    httpResponseHeaders.clear();
```
Comment (vboctor) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: Why is this no longer needed?

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java (hunk 6 above)
Comment (candrews) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: I believe we do need to use ConcurrentHashMap here. I believe that ExchangeService should represent an Exchange server, and as such, it can be used by multiple threads concurrently.

The API currently allows the user to change the httpHeaders (IMHO, this variable would be better named "httpRequestHeaders" for clarity) at any time. In other words, the httpHeaders could have entries added or removed while another thread is also adding or removing entries while another thread is actually doing requests/responses. For that reason, httpHeaders must be thread safe. In my opinion, the API right now is definitely not ideal - that's why I recommended the API be changed to use an immutable ExchangeService with a builder approach, that way, the request headers would be provided using the builder and thereafter be immutable, eliminating this complexity - but we should handle that in a separate pull request.

If a user wants to connect to multiple Exchange servers, then they'll need one instance of ExchangeService for each server.

A user could create multiple ExchangeService instances for the same Exchange server, but that's not necessary, is less efficient, and more complex to write.

So, directly replying to each point:
A single instance of ExchangeService can be used by a single server: Yes.
Multiple threads can run in parallel each using its own instance of ExchangeService: Yes
Static methods are thread safe: Yes

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java (hunk 5 above)
Comment (candrews) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: Yes, volatile is absolutely required here in order to ensure coherency between multiple threads. Please see https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java for a full explanation.

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java (hunk 2 above)
Comment (candrews) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: Okay, we'll handle the changing of "256" and "8" into constants in another issue/PR. Sounds good!

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java [hunk 9]
```diff
@@ -252,29 +233,37 @@ public void setMaximumPoolingConnections(int maximumPoolingConnections) {
...
 
   @Override
   public void close() {
-    try {
-      httpClient.close();
-    } catch (IOException e) {
-      LOG.debug(e);
+    if(httpClient!=null) {
+	  synchronized(httpClientLock){
+	    if(httpClient!=null) {
```
Comment (vbauer) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: Maybe it will be better to move this code in something like IOUtils.closeQuietly to prevent copy-paste (and cleanup other places)?

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java [hunk 10]
```diff
@@ -820,6 +805,48 @@ public void setWebProxy(WebProxy value) {
   public Map<String, String> getHttpHeaders() {
     return this.httpHeaders;
   }
+  
+  protected final CloseableHttpClient getHttpClient(){
```
Comment (vbauer) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: How to configure/use custom httpClient or poolingHttpClient (with some additional configuration)? Maybe it will be better to move some part of code in factory methods?

**Comment by vbauer** [comment]
Just a few thoughts:

* I think it would be great to have some tests to check thread-safety, otherwise it's just a guessing.
* I'm not sure that it's a good idea to design code around "synchronized" blocks if it is not necessary (but I understand that it could be a possible solution), because of theirs bottleneck-nature. Maybe it is possible to re-write code with object pool or use only prototype objects.

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java (hunk 9 above)
Comment (candrews) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: Sure, I can do that. But wouldn't it be better to do that in another pull request?

I'm happy to do it here, but to me, it seems like this pull request should solve one issue (thread safety) and not expand to other issues (like this kind of refactoring/cleanup). Let me know.

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java (hunk 10 above)
Comment (candrews) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: Sure, I can do that. But wouldn't it be better to do that in another pull request?

I'm happy to do it here, but to me, it seems like this pull request should solve one issue (thread safety) and not expand to other issues (like this one, which has already been reported separately as #423). Let me know.

**Comment by candrews** [comment]
@vbauer I don't think it's guessing - to prove thread safety, you have to do mathematical proofs. Thread safety isn't provable using unit tests. If you search the web for this thread safety testing, you'll find that this is the case - for example, http://programmers.stackexchange.com/questions/149374/testing-thread-safety explains the situation a bit.

I could add some tests, but I really don't want to get bogged down adding tests that really don't prove anything and IMHO don't add any value. I think that most people who are familiar with multithreaded programming would agree that this code is thread safe - perhaps you could ask someone familiar with multithreaded programming whom you trust to review this code instead?

Furthermore, I'm trying to work with you and get my contribution accepted for my own benefit and everyone else's - if there's some significant amount of other work you want done, please feel free to help as well - I'll do whatever I reasonably can, but I'm not going to spend an exorbitant amount of time on this issue. I already have deadlines that I'm in threat of missing so I will have to abandon this PR and fork the library at some point, I'd just very much rather not do that. So please help me help all of us by providing some clarity and help.

Regarding synchronized blocks, I completely agree with you that it's not a good idea to design code around unnecessary synchronized blocks. However, there are no such instances of unnecessary synchronized blocks here. There are 4 uses of synchronized: 2 to get the http clients (one each for pooling and non-pooling) and 2 to close the http clients (again, one each for pooling and non-pooling). In both cases, double checked locking is used so that the synchronized keyword is only hit (and therefore the lock is only acquired) only when absolutely necessary - under normal operation, there is zero locking. If you have a suggestion as to how to improve this design further (I already suggested a builder/immutable pattern, but such an API change is beyond the scope of this work), please share and if it's a reasonable effort I'll definitely implement it.

This thread safety issue is *HUGE* - the library is currently useless. Can you please clearly tell me what I need to do to get this PR accepted? So far, my understanding of the status is:
* I have to adjust some white space
* I'm not sure what you want done about the "magic number" situation with "256" and "8" - it seems beyond the scope of this PR. Can you please clarify?
* Are you happy with the ConcurrentHashMap? Or do I need to do anything for that?
* Refactor to use a common "closeQuietly" method (why is the scope of this work expanding?)
* Resolve issue #423 (again, why is the scope of the issue expanding?)

**Comment by candrews** [comment]
Any news?

Thanks! I hope you enjoy the long weekend

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java (hunk 3 above)
Comment (serious6) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: requirements are defined in [Googles Java Guide](https://google.github.io/styleguide/javaguide.html) which has been discussed to be our project reference. This is also as well mentioned in the [CONTRIBUTING.md](https://github.com/OfficeDev/ews-java-api/blob/master/CONTRIBUTING.md#coding-conventions). Using checkstyle has as well been discussed but comes along with several problems:
* usability of checkstyle-plugins (for different IDEs) differ from the checkstyle-google-styleguide
* project contains multiple formattings that where not touched by IntelliJ for some reasons these would have to be handled manually, maybe there are also bugs in the `codeStyleSettings.xml` didnt had that much time to step in.
If you have any ideas on how to do this "at low cost" feel free to discuss at #4. You are pushing at open doors with this. But whether or not we have these automated checks every PR has to match these criterias.

**Comment by serious6** [comment]
to answer your questions:
1. Matching the styleguide is necessary yes. Sorry that we dont have central checkstyle configuration but if needed there is also the possibility to run local checks. Modifying the suggested lines shouldnt be that much of work btw.
2. is already answered by @vboctor 
3. using `ConcurrentHashMap` should be ok at this point. Any follow up PR relating this topic are welcome. 
4. +1 for another PR
5. +1 for another PR

Some sidenodes:
> I think that most people who are familiar with multithreaded programming would agree that this code is thread safe - perhaps you could ask someone familiar with multithreaded programming whom you trust to review this code instead?

There is no doubt that you will have tested your code. In general the reason to add unit-tests is to guarantee quality also over the next changes that will possibly be done by others. Regarding your point of a 'trusted person' we so far put our PRs under review by the community so everyone is able to comment. And as you might have noticed there is some related feedback improving the quality of this PR. To solve this I would suggest to include your fix in the next version. So we will have this merged after 2.0 has been released. So maybe we will be able to also include some follow up PR on this one. Any comments?

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java (hunk 6 above)
Comment (vboctor) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: I had a typo that was critical here:
"A single instance of ExchangeService can be used by a single server."  I meant a single instance is used by a single thread.  i.e. ExchangeService is not thread safe, but it is OK to have 3 threads using 3 different instances of ExchangeService connecting to same or difference servers in parallel.

It seems that you want to have a single instance of ExchangeService used by multiple threads in parallel.  What is the motivation for that.  That adds complexity and will have weird cases if the state can depend on the request (e.g. if headers may need to change based on request for example).

> I believe we do need to use ConcurrentHashMap here. I believe that ExchangeService should represent
> an Exchange server, and as such, it can be used by multiple threads concurrently.

If we go with the model that I'm suggesting, then that we don't need this protection against concurrency.

**Comment by vboctor** [comment]
I'm still not sold on the need for this PR, so I added added comment with my question and explaining my point.  I may be convinced though.

File: src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java (hunk 6 above)
Comment (candrews) [inline src/main/java/microsoft/exchange/webservices/data/core/ExchangeServiceBase.java]: The way ExchangeService is implemented today it isn't thread safe even when you create new instances of it for each thread - see #371.

If it is thread safe (as this PR makes it), ExchangeService becomes:
# Easier to use. I believe developers would assume it's thread safe, as *Service named classes usually are. It's also less code to use it thread safe rather than creating, configuration, then closing it for each use.
# Faster. Before this PR, the suggested proper use of ExchangeService is to create and then tears down an HttpClient every time it's used - that takes time. It's also unable to share HttpClient resources (like thread pools) as the HttpClient design intends, reducing performance.
# Less memory intensive. By having fewer instances of ExchangeService, less memory is used. Most significantly, by sharing HttpClient instances, thread pools, caches, and objects are all shared (as HttpClient is intended to be used).

**Comment by candrews** [comment]
1. You're right in that the whitespace changes aren't a big deal. I've just added a commit (04a3200806429057297f8c1c379c4cfe05411f36) addressing them and ensured that checkstyle passes now. If you'd like, I could rebase/merge this PR's now 2 commits into one - just let me know.
In order to prevent myself (and others!) from causing whitespace headaches for you going forward, I submitted a PR adding checkstyle to the build: #438
2. :+1: 
3. :+1: 
4. I submitted PR for that: #439 
5. Managing the scope of a shared HttpClient between ExchangeService instances seems like it'll be tricky: the current API hides the httpClient from the user, including closing it when ExchangeService is closed. If we allow the user to specify an httpClient, then should ExchangeService close it when it's done or not? And if the user doesn't specify one, should ExchangeService close it? I don't think it's insurmountable, but I'll leave that for later or for someone else to tackle. Ideally, I think a builder for ExchangeService with an immutable ExchangeService API should be created, and the specifiable httpClient feature can be added at the same time.

I also replied to @vboctor's comment/concerns at https://github.com/OfficeDev/ews-java-api/pull/428#discussion_r38964148

Thanks again

**Comment by codewheeney** [comment]
Are there any updates for this PR?  We're actively blocked by this issue at the moment.

**Comment by codewheeney** [comment]
I cloned off of candrews fix and tested, and this solution is necessary but not sufficient to fix the concurrency issue I'm seeing.  There appears to be a bug in Apache's CloseableHttpClient, namely that, although it's marked as thread safe, it uses the not threadsafe NTLMScheme class.  Concurrent access can cause out of bounds exceptions when writing to internal buffers due to static objects in NTLMEngineImpl. 

I've opened this issue with apache:  https://issues.apache.org/jira/browse/HTTPCLIENT-1686

**Comment by codewheeney** [comment]
The issue I reported about HttpClient has been closed as invalid, although I think the maintainer of that code doesn't quite see my point, unless I'm not following what this PR is attempting to do.  Fundamentally, this PR does the right thing by creating a separate HttpClientContext for each concurrent access.  That should be sufficient according to the comments in HTTPCLIENT-1686, but it isn't in my testing because of the static field in NTLMEngineImpl.java.

I'd appreciate any other insight into testing or use of this code in this PR that others are seeing.

**Comment by schakko** [comment]
@candrews Thank you for this pull request, but it doesn't solve all concurrency problems: I am stumbling upon the exactly same problem @codewheeney described.

In our case we have two threads (one HTTP worker thread, one background thread) which are accessing the Exchange web service through multiple instances of ExchangeService. 
HTTP worker thread pulls randomly data from the Exchange web service, the background check pulls data every two minutes. If the requests are occuring at the same time, the exception

	Caused by: java.lang.ArrayIndexOutOfBoundsException: 40
		at org.apache.http.impl.auth.NTLMEngineImpl$NTLMMessage.addByte(NTLMEngineImpl.java:911)
		at org.apache.http.impl.auth.NTLMEngineImpl$NTLMMessage.addULong(NTLMEngineImpl.java:940)
		at org.apache.http.impl.auth.NTLMEngineImpl$Type1Message.getResponse(NTLMEngineImpl.java:1043)

occurs. It doesn't matter if I use the same instance of ExchangeService, create multiple instances for both threads or set the maximumPoolingConnections to 1 in any of those combinations.

At the moment, @codewheeney suggestion to modify the HttpClient seems to be the only working solution - which I do want to avoid b/c of obvious reasons.
Are there any other ideas or did you made progress in this case, @codewheeney ?

**Comment by codewheeney** [comment]
I have not re-opened https://issues.apache.org/jira/browse/HTTPCLIENT-1686 yet.  At the moment, I've forked HTTP Client and EWS into our repository and I'm working from patched versions. 

It'd be great if @schakko commented on https://issues.apache.org/jira/browse/HTTPCLIENT-1686 and re-opened it, pointing out that the static data member in NTLMEngineImpl will never work properly in a multithreaded environment.  I failed to get that point across and haven't had cycles to get in there and re-argue the case.

**Comment by schakko** [comment]
@codewheeney I left a comment on issues.apache.org but it's hard to argue if they don't want to make it thread-safe (and more reliable).
BTW: I fixed the EWS API without touching the HTTP client. Feel free to pull my branch from https://github.com/schakko/ews-java-api/tree/gh-371-make-ews-thread-safe.

**Comment by wwahmed** [comment]
Hello @schakko 

I see the changes you have made to make the API more threadsafe. However, I am confused about your comment that you have done so without touching the HTTPClient when the fork you are on, had the commit #5541396 from lamielle which indicates that it's being built with the forked httpclient (httpclient-netc, upgrade to 4.5.1) and it's still present in pom.xml in your commit. 

Do the changes in your fork require everything done previously by candrews and lamielle - including the httpclient update?

Thanks for your help.

**Comment by schakko** [comment]
@wwahmed I understand your confusion.
I compiled and (unit-)tested it against 4.4.1 and it did the job. But I must admit that I did not test it in production.

Our working ews-java-api has the following history:
- 690c75e: was still in our local repository, pushed it a few seconds ago; does only change httpclient-netc to httpclient and removes the custom repository. Revokes 55413968488
- 1903b73aea2: my fix, does compile/test against 4.4.1
- 7cca414b32b: obsolete; no code changes
- 55413968488: introduces httpclient-netc-4.5.1, revoked through 690c75e
- 66b107b9c25: coding standards
- 0020b636eb0: initial work for this issue against httpclient 4.4.1

The build with this history has been running in production for 3 months without any problems. In this time the API has been accessed by ~8-10 threads concurrently all the time.

**Comment by wwahmed** [comment]
Hi @schakko  Thanks for confirming - I have also tested it without the httpclient-netc-4.5.1 http client and it does work with 8-10 concurrent threads. 

Thanks a lot for your confirmation.

**Comment by serious6** [comment]
@schakko maybe you can share the unit tests you created? Or provide a PR in this? Increasing the coverage might ve useful.

**Comment by SebastianWe** [comment]
When using multiple threads a "ServiceRequestException: The request failed. Connection is still allocated" might occure.

I added a fix for this in #491.
